src/constraints.jl

```diff
@@ -0,0 +1,207 @@
+#############################################################################
+# AbstractShape
+# Abstract base type for the shape of constraints
+"""
+    AbstractShape
+
+Abstract vectorizable shape. Given a flat vector form of an object of shape
+`shape`, the original object can be obtained by [`reshape`](@ref).
+"""
+abstract type AbstractShape end
+
+"""
+    dual_shape(shape::AbstractShape)::AbstractShape
+
+Returns the shape of the dual space of the space of objects of shape `shape`. By
+default, the `dual_shape` of a shape is itself. See the examples section below
+for an example for which this is not the case.
+
+## Examples
+
+Consider polynomial constraints for which the dual is moment constraints and
+moment constraints for which the dual is polynomial constraints. Shapes for
+polynomials can be defined as follows:
+```julia
+struct Polynomial
+    coefficients::Vector{Float64}
+    monomials::Vector{Monomial}
+end
+struct PolynomialShape <: JuMP.AbstractShape
+    monomials::Vector{Monomial}
+end
+JuMP.reshape(x::Vector, shape::PolynomialShape) = Polynomial(x, shape.monomials)
+```
+and a shape for moments can be defined as follows:
+```julia
+struct Moments
+    coefficients::Vector{Float64}
+    monomials::Vector{Monomial}
+end
+struct MomentsShape <: JuMP.AbstractShape
+    monomials::Vector{Monomial}
+end
+JuMP.reshape(x::Vector, shape::MomentsShape) = Moments(x, shape.monomials)
+```
+The `dual_shape` allows to define the shape of the dual of polynomial and moment
+constraints:
+```julia
+dual_shape(shape::PolynomialShape) = MomentsShape(shape.monomials)
+dual_shape(shape::MomentsShape) = PolynomialShape(shape.monomials)
+```
+"""
+dual_shape(shape::AbstractShape) = shape
+
+"""
+    reshape(vectorized_form::Vector, shape::AbstractShape)
+
+Return an object in it original shape `shape` given its vectorized form
+`vectorized_form`.
+
+## Examples
+
+Given a [`SymmetricMatrixShape`](@ref) of vectorized form `[1, 2, 3]`, the
+following code retrieve the matrix `Symmetric(Matrix[1 2; 2 3])`:
+```julia
+reshape([1, 2, 3], SymmetricMatrixShape(2))
+```
+"""
+function reshape end
+
+"""
+    shape(c::AbstractConstraint)::AbstractShape
+
+Return the shape of the constraint `c`.
+"""
+function shape end
+
+"""
+    ScalarShape
+
+Shape of scalar constraints.
+"""
+struct ScalarShape <: AbstractShape end
+reshape(α, ::ScalarShape) = α
+
+"""
+    VectorShape
+
+Vector for which the vectorized form corresponds exactly to the vector given.
+"""
+struct VectorShape <: AbstractShape end
+reshape(vectorized_form, ::VectorShape) = vectorized_form
+
+##########################################################################
+# Constraint Reference
+# Holds the index of a constraint in a Model.
+# TODO: Rename "m" field (breaks style guidelines).
+struct ConstraintRef{M <: AbstractModel, C, Shape <: AbstractShape}
```

Comment:
Style: are we putting spaces around `<:`?

Reply:
I thought so as it is a binary operator but we might make an exception for this one